glean-core/ffi/src/lib.rs

```diff
@@ -342,23 +342,58 @@ pub extern "C" fn glean_is_first_run() -> u8 {
     with_glean_value(|glean| glean.is_first_run())
 }
 
+// Unfortunately, the way we use CFFI in Python ("out-of-line", "ABI mode") does not
+// allow return values to be `union`s, so we need to use an output parameter instead of
+// a return value to get the task. The output data will be consumed and freed by the
+// `glean_process_ping_upload_response` below.
+//
+// Arguments:
+//
+// * `result`: the object the output task will be written to.
 #[no_mangle]
-pub extern "C" fn glean_get_upload_task() -> FfiPingUploadTask {
-    with_glean_value(|glean| FfiPingUploadTask::from(glean.get_upload_task()))
+pub extern "C" fn glean_get_upload_task(result: *mut FfiPingUploadTask) {
+    with_glean_value(|glean| {
+        let ffi_task = FfiPingUploadTask::from(glean.get_upload_task());
+        unsafe {
+            std::ptr::write(result, ffi_task);
+        }
+    });
 }
 
-// We need to pass the whole task instead of only the document id,
-// so that we can free the strings properly on Drop.
+/// Process and free a `FfiPingUploadTask`.
+///
+/// We need to pass the whole task instead of only the document id,
+/// so that we can free the strings properly on Drop.
+///
+/// The data behind the `task` pointer should be freed after return by the caller.
```

Comment:
Reading this now again I thin we should slightly change this:

```suggestion
/// After retrun the `task` should not be used further by the caller. 
/// Any data from the `UploadRequest` has been freed.
```

(not 100% happy with that phrasing either).
I just want to call out that while the Rust side frees the contained strings (if any), the other side need to remove the variable (at best it's just on the stack, so nothing to do anyway)